### PR TITLE
Multiproject plate: notebooks changes!

### PR DESCRIPTION
**Pull request by AmandaBirmingham**
Ok, this one gets us to the end-game: notebook changes!  It includes:

* a reorganized matrix tube notebook that (a) takes in a lot of the inputs right up front, (b) has users provide study information that is *separate* from plate information, since now plates can have more than one study, (c), supports loading in multiple sample accession files, metadata files, counts files for the multiple studies, and (d) works out the potentially-multiproject control info to send to the sample sheet creation.  It also addresses Gail's request that wet lab users be able to input the RKL tracking number as experiment name. (These various changes also required some changes to the test files to exercise cases with multiple projects on a plate.)
* several "support" functions in the python library; the main goal of these is to prevent users either from having to enter the same data in multiple places (e.g., qiita ids, project abbreviations, etc) or from having to merge information manually.
* new functionality within `compress_plates` to name compressed plates according to the rules agreed on with the wet lab in cases where more than one project is included on the compressed plate.  
* a new `literals` module for frequently-repeated constants. I realize this is widely considered an antipattern and I have explained in some detail why I went this route in the code comments.  My aim is for this to be a temporary measure while we tease apart all the repeated string usages.  
* refactoring of the code to extract a qiita id from the end of a project name, which was doing the same string handling in quite a few different places.

This also includes the new work work added to master from Charlie to pull override cycles from the run info and allow them to be set in the sample sheet.

NOTE: What this does NOT support yet is multiproject plates for *absolute quant*.  This is a big enough chunk that I didn't want to throw that in too, but it is coming next ... yet *more* changes to `sample_sheet` :). It also doesn't support multiproject plates for transcriptomics yet.

### CODE REVIEW

File: metapool/literals.py
Comment (antgonza) [inline metapool/literals.py:33]: IMOO, if collision is a possibility, I prefer to use something more complex, normally with some prefix, like: 'metapool_samples'

File: metapool/literals.py
Comment (antgonza) [inline metapool/literals.py:51]: Out of curiosity, why have a function for this?

File: metapool/literals.py
Comment (antgonza) [inline metapool/literals.py:88]: Same here, why a function?

File: metapool/literals.py
Comment (antgonza) [inline metapool/literals.py:96]: Why [1]?  ... BTW `_split_plate_name` is used in 2 places: one for [0] and another [1]. For these kind of scenarios, I think is better to remove the function/method and just do that code twice. However, I understand that this is a personal preference.

File: metapool/scripts/seqpro.py
Comment (antgonza) [inline metapool/scripts/seqpro.py:78]: This line will only be ran if we are in verbose mode, right? Should it be out of the if?

File: metapool/scripts/seqpro_mf.py
Comment (antgonza) [inline metapool/scripts/seqpro_mf.py:68]: Same here, line is in if, should it be out?

File: metapool/tests/test_metapool.py
Comment (antgonza) [inline metapool/tests/test_metapool.py:1382]: This will be printed _all_ the time, not only when assertion fails, right? BTW I think that when a tests fails, it shows you why it failed and the message will have the values.

File: metapool/tests/test_metapool.py
Comment (antgonza) [inline metapool/tests/test_metapool.py:1388]: Not sure if this print is needed.

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (antgonza) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:62]: Should we have the total of steps here? I mean something like Part 1/X. 

EDIT: I think having both the parts and the steps will be helpful, sometime like Part 1/4, Step 1/X.

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (antgonza) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:122]: OK to leave this email here?

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (antgonza) [inline notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:254]: This will "print" the shape but not sure if we actually need that and/or if we should have an actual print or metadata.df.head() or something like that. 

Edit: BTW this applies to all cells where the last line is XX.shape.

**Review by antgonza — COMMENTED** [review]
Some questions and comments. Thank you for working on this @AmandaBirmingham .

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py:51]: That is true, getters() are IMHO less pythonic than the @property decorator. I tend to just directly use the variable since all members are public in Python but I believe the clean way is to use a property decorator.

File: metapool/tests/test_metapool.py
Comment (AmandaBirmingham) [reply metapool/tests/test_metapool.py:1382]: @antgonza , this is @charles-cowart's new functionality that you merged into master on Wednesday: https://github.com/biocore/metagenomics_pooling_notebook/blob/95e78dcd16342967a6ef73666cee9b1a5e27969d/metapool/tests/test_metapool.py#L1289 .  All I did was back-merge it from master into the development branch.

File: metapool/tests/test_metapool.py
Comment (AmandaBirmingham) [reply metapool/tests/test_metapool.py:1388]: Same situation as above.  Perhaps @charles-cowart can weigh in?

File: metapool/tests/test_metapool.py
Comment (charles-cowart) [reply metapool/tests/test_metapool.py:1382]: It will be printed all the time to stdout, but it will only be shown if the test fails, in which case the fp will be useful for debugging.

File: metapool/tests/test_metapool.py
Comment (charles-cowart) [reply metapool/tests/test_metapool.py:1388]: It will be printed all the time to stdout, but it will only be shown if the test fails, in which case the fp will be useful for debugging.

File: metapool/tests/test_metapool.py
Comment (charles-cowart) [reply metapool/tests/test_metapool.py:1382]: I discussed it with Antonio and I mentioned I planned to revisit the logging library in the near future. I'll go ahead and issue a PR to remove the print statements in master immediately and you can merge the lastest master into this branch afterward or I can remove them after this branch has been merged. Whatever sounds easiest for everyone.

File: metapool/tests/test_metapool.py
Comment (charles-cowart) [reply metapool/tests/test_metapool.py:1388]: I discussed it with Antonio and I mentioned I planned to revisit the logging library in the near future. I'll go ahead and issue a PR to remove the print statements in master immediately and you can merge the lastest master into this branch afterward or I can remove them after this branch has been merged. Whatever sounds easiest for everyone.

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:33]: Good idea, will do.

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [reply notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:254]: Yes, it is intended to print the shape.  Experience using the matrix tube notebook with the lab has shown that printing out the dataframe (or using head()) is insufficient to check that the expected dataframe was accurately created, because those display methods cut off rows and columns.  Shape makes it easy to see in an instant, e.g., "Oh, something must be wrong because I added two 10 line files but ended up with a dataframe with only 10 records" :)

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [reply notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:122]: Real emails are all over the notebooks and test files, but you have a good point--I shouldn't perpetuate that!  I will replace it with a dummy.

File: notebooks/matrix_tube_pipeline_seqcount_norm.ipynb
Comment (AmandaBirmingham) [reply notebooks/matrix_tube_pipeline_seqcount_norm.ipynb:62]: Ok, I can add something to make that clearer.  In the long run, my plan is to break this one gigantic notebook into multiple notebooks, for the reasons discussed in FutureTech; once we get there, the different parts will no longer be an issue.

File: metapool/tests/test_metapool.py
Comment (AmandaBirmingham) [reply metapool/tests/test_metapool.py:1382]: @charles-cowart I say go ahead and change them :)

File: metapool/scripts/seqpro.py
Comment (AmandaBirmingham) [reply metapool/scripts/seqpro.py:78]: @antgonza again, I cannot speak to that.  The `get_qiita_id_from_project` call is just a drop-in replacement for the original string `replace` call; I explicitly did NOT change any logic.  I'm not sure who wrote this module, but I suspect @charles-cowart  is the expert here.

File: metapool/scripts/seqpro_mf.py
Comment (AmandaBirmingham) [reply metapool/scripts/seqpro_mf.py:68]: Same comment as above :)

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:88]: Because calling this function replaces doing in-the-weeds string manipulation *all over* this code base, which is a very brittle idiom.  Right now functions spread across three repositories have baked-in knowledge of the string format of project names, and this means that if we ever need to change anything about that format--or even if we discover we need to be more careful in interpreting the *existing* format (as has happened more than once with the string handling used to parse fastq file names), we have a big, fiddly task to find and modify all those functions (and we're pretty likely to miss one somewhere ... ).  We should take a lesson from the situation with the blanks, where the identification of blanks via string matching was happening separately in five places across three repos--but was only the same in *four* of them, and the outlier was incompatible with the new NPH project.

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:96]: I'll refer here to my comment above about string handling.  I feel really strongly about this one.

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:51]: I believe that an  @property decorator cannot be used here because there is no class for the property to be attached to.  I am willing to take this out.

File: metapool/literals.py
Comment (antgonza) [reply metapool/literals.py:84]: Sounds good, thank you for the explanation. I guess this highlights the "advantages" of OOP, where we could have a class "Project", then a property/method `name` and another `qiita_id`, which are only parsed once and used along the code and inherited by other software/code vs. multiple smaller functions. Note that I'm not suggesting to make this change.

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py]: > I believe that an @Property decorator cannot be used here because there is no class for the property to be attached to. I am willing to take this out.

Good point 😆  Since it's not in a class using the literal directly seems cleaner.

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py:84]: one of the few yes? 😄

File: metapool/scripts/seqpro.py
Comment (charles-cowart) [reply metapool/scripts/seqpro.py:78]: no for this particular case verbose is intended. I do have a PR waiting for review that addresses seqpro but it's awaiting review. ;)

File: metapool/scripts/seqpro_mf.py
Comment (charles-cowart) [reply metapool/scripts/seqpro_mf.py:68]: Ditto^2 :) :)

**Review by charles-cowart — CHANGES_REQUESTED** [review]
Looks good, thank you, and thank you for your patience! I just have a couple of concerns.

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:4]: For my edification, how worried are we about recompile time?

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:49]: Hmm, I agree that it is always hard to come up with good nomenclature.  All the contents of this module--both the constants and the functions--are centered around *string names*.  The constants codify some string names that are used in various places throughout the code, and the functions encode knowledge of the naming conventions for other string names used throughout the code (and perform string operations on these names based on the encoded knowledge).  This allows code outside this module to e.g. break apart the name of a project into its component pieces *without* having to have any knowledge of how the pieces are actually represented in the name, because both the knowledge of the expected format and the constants defining expected delimiters are all bundled together in here.  I am happy to change the name of this module to something that would better represent its mission (and very open to suggestions for what that would be ... :D )

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:76]: I think we are on the same page here! I went through `metapool` and thought I had found most instances of project name mining, but I am intensely interested in any I missed! :) I also went through `mg-scripts`, where I found `Pipeline._parse_project_name()` and  (`ConvertJob._get_sample_sheet_info` in the new `copy_sequences` branch).  I have a working branch of `mg-scripts` that modifies these to use the centralized `literals.parse_project_name()` function which, as you suggest, creates a dictionary of k/v pairs.

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:53]: This is very good to know!  One of these days (but probably not today :D ), it would be nice to get them using a single, shared definition, which shouldn't be a big modification: several modules of `qp_klp` already import the `metapool` module, so hopefully this change will just be importing a couple of additional constants :)

File: metapool/tests/test_sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/tests/test_sample_sheet.py:1578]: Good idea!  I will add that!

File: metapool/notebook_support.py
Comment (AmandaBirmingham) [reply metapool/notebook_support.py:11]: Hmm, yeah, I guess I *could* just do that, couldn't I? Thanks! :D

File: metapool/notebook_support.py
Comment (AmandaBirmingham) [reply metapool/notebook_support.py:167]: Argh, nomenclature rears its ugly head again :D  Here's the thing: although the *repo* is named `metagenomics_pooling_notebook`, that in fact seems to be a misnomer.  The `metapool` module is used all over, in parts of the code that have (AFAICT) no notebook interaction, like `qp_klp.Step`, `qp_klp.klp`, `mg-scripts.Pipeline`, `mg-scripts.ConvertJob`, etc.  I realize that `notebook_support` is a filthy cop-out (as is `util` :D ) ... I kind of like `df_util` because it sounds more specific, but since only 2 of the 3 public functions in this module actually act on dfs (`extend_compression_layout_info` acts only on dictionaries), it is somewhat misleading specificity.  I'm willing to change this to `util`, or discuss other names if you prefer ...

File: notebooks/test_output/Indices/YYYY_MM_DD_Celeste_Adaptation_16-21_indices.txt
Comment (AmandaBirmingham) [reply notebooks/test_output/Indices/YYYY_MM_DD_Celeste_Adaptation_16-21_indices.txt:674]: Ummm, I am lost here, and I think I am going to need you to walk me through what you are seeing in more detail.  To me it appears the only differences between the file in the PR and the file in master is in the `Sample` column, where some of the names have changed.  It looks to me like every sample has two different entries, one for the iTru5_plate and one for the iTru7_plate; in the example you highlighted, the name `46L.Month.6.1` has been replaced by `20001CC.B.FW2`, but other than that, the index assignments haven't changed since `20001CC.B.FW2` has the same iTru5 and iTru7 indices as `46L.Month.6.1` used to:

Screenshot 2024-08-21 at 3 23 49 PM

<p />
<p />

Screenshot 2024-08-21 at 3 24 07 PM

Can you give me more clarity on what you're seeing?

File: metapool/sample_sheet.py
Comment (AmandaBirmingham) [reply metapool/sample_sheet.py:1766]: My approach in this refactor was to NOT remove **any** existing functions. Thus, every existing function call, no matter where it may be buried in the various test or repos (which I am definitely not familiar with all of!), should continue to work, even if the work is now just being forwarded to a new function.  Basically, I'm trying not to rock the boat *too* much.  Given this, I didn't remove or rename `sheet_needs_demuxing`.  

In the future, we might change this.  We would need to take into account that `sheet_needs_demuxing` is tolerant of the lack of CONTAINS_REPLICATES_KEY, whereas `_get_contains_replicates_value` (and the other place outside of `sheet_needs_demuxing` where `_get_contains_replicates_value` is used) are NOT tolerant of that and need to throw errors.  Still, if we definitely want to combine the two, I'm sure we could find a way to switch on this.

I used the `_` to indicate the new function I added is supposed to be private to this module.  I did this because right now it *isn't* being used anywhere outside this module, and I think it should stay that way until we decide if we want ;)

File: metapool/tests/test_metapool.py
Comment (charles-cowart) [reply metapool/tests/test_metapool.py]: I took the liberty of pushing directly to this branch, my apologies. The statements are removed! :)

File: metapool/tests/test_metapool.py
Comment (charles-cowart) [reply metapool/tests/test_metapool.py]: I took the liberty of pushing directly to this branch, my apologies. The statements are removed! :)

File: metapool/tests/test_sample_sheet.py
Comment (charles-cowart) [reply metapool/tests/test_sample_sheet.py:1583]: ty!

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py:4]: I'm not too worried that it will drastically impact recompile times; however if it did that would impact developing and running tests, which wouldn't be great. I haven't seen this antipattern used in a while so I'm curious to see if it has any real-world effect!

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py:49]: Hmm, in my experience this type of file is usually called constants.py and doesn't contain functions. However that's not really any more descriptive than literals.py in this case.

Although 'strings' is not a reserved word, IMHO it's too close to the built-in 'string' module. The string module also has helper functions though. Maybe something analogous like mpstrings.py, mp_strings.py, MPStrings.py?

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py:76]: That sounds great, thanks! So we would use literals.parse_project_name() for everything and remove get_short_name_and_id() and whatnot?

File: metapool/notebook_support.py
Comment (charles-cowart) [reply metapool/notebook_support.py:163]: 😄 Sorry, I know over time it grows increasingly misnamed. 😄 But not adding to that is good, right? I can go with util.py if you're okay with it.

File: metapool/sample_sheet.py
Comment (charles-cowart) [reply metapool/sample_sheet.py:1766]: Not removing existing functions in this PR is reasonable. Yeah I think we should combine the two at some point.

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py:76]: Note that this function is discussed in other conversations below.

File: notebooks/test_output/Indices/YYYY_MM_DD_Celeste_Adaptation_16-21_indices.txt
Comment (charles-cowart) [reply notebooks/test_output/Indices/YYYY_MM_DD_Celeste_Adaptation_16-21_indices.txt:674]: Ah ty! Yes, you are right. Once I pulled both files out and reviewed them it all became clear.

File: metapool/literals.py
Comment (AmandaBirmingham) [reply metapool/literals.py:49]: sold!

File: metapool/notebook_support.py
Comment (AmandaBirmingham) [reply metapool/notebook_support.py:163]: done

Comment (charles-cowart) [reply metapool/notebook_support.py]: ty!

File: metapool/literals.py
Comment (charles-cowart) [reply metapool/literals.py]: ty!

**Review by charles-cowart — APPROVED** [review]
Everything looks good, thank you very much!

**Review by antgonza — APPROVED** [review]
Looks good, thank you!

**Pull request merged**
